### PR TITLE
OI-165: Charts filters

### PR DESCRIPTION
**Pull request by nvelychenko**

### CODE REVIEW

File: modules/openideal_statistics/openideal_statistics.routing.yml [hunk 1]
```diff
@@ -5,3 +5,11 @@ openideal_statistics.graphs_reports:
     _title: 'Graphs & Reports'
   requirements:
     _permission: 'access content'
+
+openideal_statistics.openideal_statistics_date_select_form:
```
Comment (db-Matroskeen) [inline modules/openideal_statistics/openideal_statistics.routing.yml]: Do we need a separate page for the form?

File: modules/openideal_statistics/src/Form/OpenidealStatisticsDateSelectForm.php [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+  const DATE_TYPE_FIXED = 'fixed';
+  const FROM = 'from';
+  const TO = 'to';
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container) {
+    $instance = parent::create($container);
+    $instance->getRequest();
```
Comment (db-Matroskeen) [inline modules/openideal_statistics/src/Form/OpenidealStatisticsDateSelectForm.php]: Looks weird. Why do we need this?

File: modules/openideal_statistics/src/Form/OpenidealStatisticsDateSelectForm.php [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+  /**
+   * {@inheritdoc}
+   */
+  public function validateForm(array &$form, FormStateInterface $form_state) {
+    $values = $form_state->getValues();
+    if ($values[self::DATE_TYPE] == self::DATE_TYPE_FIXED && !$values[self::FIXED_RANGE]) {
+      $form_state->setErrorByName(self::FIXED_RANGE, $this->t('An option should be selected'));
+    }
+    elseif ($values[self::DATE_TYPE] == self::DATE_TYPE_CUSTOM && (!$values[self::FROM] && !$values[self::TO])) {
+      $form_state->setErrorByName('', $this->t('You must specify both dates'));
```
Comment (db-Matroskeen) [inline modules/openideal_statistics/src/Form/OpenidealStatisticsDateSelectForm.php]: Should we specify element name?

File: modules/openideal_statistics/src/Plugin/Block/OpenidealStatisticsEntityByChart.php [hunk 4]
```diff
@@ -141,4 +123,47 @@ public function blockSubmit($form, FormStateInterface $form_state) {
...
+      $query->condition('created', $filters[OpenidealStatisticsDateSelectForm::FROM], '>=');
+    }
+
+    if ($entity_by == 'category') {
+      $query->exists('field_category');
+    }
+
+    $ids = $query->execute();
+
+    $ideas = $storage->loadMultiple($ids);
```
Comment (db-Matroskeen) [inline modules/openideal_statistics/src/Plugin/Block/OpenidealStatisticsEntityByChart.php:151]: I think we may have different entity types and bundles here (not only ideas).

**Review by db-Matroskeen — CHANGES_REQUESTED** [review]

File: modules/openideal_statistics/src/Plugin/Block/OpenidealStatisticsEntityByChart.php (hunk 4 above)
Comment (nvelychenko) [reply modules/openideal_statistics/src/Plugin/Block/OpenidealStatisticsEntityByChart.php:151]: Only Ideas were requested.

**Pull request merged**